Visualization settings query processor middleware

noahmoss opened the pull request:
This PR is a draft of a new `visualization-settings` QP middleware which fetches, combines and orders visualization settings for a query. This has become unexpectedly complex and I've been playing whack-a-mole this week trying to handle different issues and edge cases I keep discovering, so I figured I'd get some feedback on what I have so far before I sink a ton of more time into it.

Ultimately, the goal is to make the _complete_ viz settings for each column available in `streaming-rff` in `metabase.query-processor.streaming` so that they can be used to format the export.

For every query, we need to have access to three pieces of information: the card-level viz settings, the base field viz settings (defined in the data model), and the order of columns in the output.

High level overview:
* For queries on **saved cards**: card-level viz settings are fetched from the DB, and combined with the base settings each field which are fetched from the QP store. They are ordered by the contents of the `:fields` key on the query.

* For queries on **unsaved cards**: viz settings are passed from the frontend in the API call, and put into the query map. The middleware pulls them out of the query map, fetches field settings from the QP store, and order them by the `:fields` key (the same as for saved cards).

* **Native query**: A viz settings map is passed from the frontend that includes both a `table.columns` key, with the order and names of columns, and `column_settings`, containing the viz settings. These can be correlated by the column name. This even seems to work if you select the same column multiple times and format them differently, since the frontend seems to makes them unique -- something like this (ignoring unrelated data):

```
{
   "table.columns":[
      {
         "name":"ID"
      },
      {
         "name":"ID_2"
      }
   ],
   "column_settings":{
      "[\"name\",\"ID\"]":{
         "show_mini_bar":true
      },
      "[\"name\",\"ID_2\"]":{
         "scale":2
      }
   }
}
```

(Field settings seem to be ignored on the frontend for native queries, so I'm assuming we should ignore them here too.) 

**Update**: I fixed the native query weirdness detailed below

* ~**Native query** weirdness: `table.columns` is not included in the viz settings on the frontend _until_ we reorder some fields. So if we run a simple query like `select id, name, email, created_at from people;` and then change a viz setting on the `created_at` column, the viz settings JSON will look like this:~

```
{"column_settings":{"[\"name\",\"CREATED_AT\"]":{"date_abbreviate":true}}}
```

~There's no way to tell which row of the output these settings should apply to. Strangely, once you've reordered anything, all subsequent queries will include `table.columns`, even if you switch back to the original ordering or even change the table/database.~

~This issue could be fixed on the frontend to always include `table.columns` in viz settings, but I'm not sure what the scope of that change is.~

jeff303 commented on src/metabase/api/common.clj:
```diff
@@ -424,9 +424,7 @@
-    (check (not (and offset (not limit))) [400 (tru "When including an offset, a limit must also be included.")])
-    ))
-
+    (check (not (and offset (not limit))) [400 (tru "When including an offset, a limit must also be included.")])))
```
Seems like an unneeded change?

camsaul replied:
it looks like it's just a fix for hanging parens. Not _needed_, but I think it makes sense to just fix these things when we come across them rather than opening a whole separate PR to fix them. If we have to open a whole new PR to fix stuff like this it's just not going to get fixed

noahmoss replied:
I don't even remember changing this tbh. I think it was fixed automatically by my editor while I was snooping around in this file.

jeff303 commented on src/metabase/api/dataset.clj:
```diff
@@ -83,25 +84,37 @@
+   export-format          ExportFormat}
+  (let [query        (json/parse-string query keyword)
+        viz-settings (mb.viz/db->norm (json/parse-string visualization_settings parse-viz-setting-key))
+        query (-> (assoc query
```
nit: line up right hand side in let, or don't, but this mixes both.

jeff303 commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+            (qp query rff' context))
+
+          (= (:type query) :native)
+          (let [columns                  (-> viz-settings :table.columns)
```
nit: line up right hand side in let, or don't, but this mixes both.

jeff303 reviewed (approved):
Backend pieces LGTM, with some nitpicky stuff.  I like how this handles the ad-hoc query case but contains the specific logic to one place.

camsaul commented on src/metabase/api/dataset.clj:
```diff
@@ -83,25 +84,37 @@
      (api/defendpoint POST [\"/:export-format\", :export-format export-format-regex]"
   (re-pattern (str "(" (str/join "|" (map u/qualified-name (qp.streaming/export-formats))) ")")))
 
+(def ^:private column-ref-regex #"\[\"ref\",\[(?:\"field\",\d+,null|\"expression\",\".+\")\]\]")
```
This seems like it could be a little too strict... the column ref could be something like 

```clj
[:ref [:field 1 {:temporal-unit :month}]]
```

(serialized as JSON) which wouldn't match this regex. Maybe just check if starts and ends with square brackets? Pretty sure that will never be a raw column name.


wouldn't it be easier to parse the JSON column key than to try

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+         (->> id
+              qp.store/field
+              :settings
+              (#(normalize-field-settings id %))
```
You could just do this one as `(normalize-field-settings id)`. `->>` threads thru the last arg. So it will do the same thing

camsaul replied:
This would maybe be a little clearer with a `for` loop as well

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+(defn- update-native-query-viz-settings
+  "For each column, select its viz settings from the column-viz-settings map by its name."
+  [columns column-viz-settings]
+  (map (fn [column] (-> column
+                        :name
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
```
This sort of stuff would probably be more clear if you wrote it using `for`. Generally I think it's preferable to use higher-level looping constructs like `for` over `map` and threading macros, especially threading macros that work around the threading using inline `#()` functions

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+    ;; Unsaved card -> viz settings passed in query
+    (-> query :viz-settings)))
+
+(defn- is-export-context?
+  [context]
+  (contains? #{:json-download :csv-download :xlsx-download} context))
```
You could write this a little more generally by doing something like

```clj
(def export-contexts
  #{:json-download :csv-download :xlsx-download})
```

and then you can just do `(export-contexts context)` where you need it... I think that's probably defined somewhere else tho IIRC. Better if we can avoid duplicating the export contexts like so if we add a new one in the future we don't have to update a ton of different parts of the code.

Also, I think `export-context?` would be a little more idiomatic of a name for a predicate function like this -- e.g. Clojure core has `string?` and `number?` rather than `is-string?` and `is-number?`. Not that important tho

noahmoss replied:
Hmm, there's `export-formats` in `metabase.query-processor.streaming` but that returns `#{:csv :api :xlsx :json}`

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+      (let [viz-settings        (get-viz-settings query)
+            column-viz-settings (::mb.viz/column-settings viz-settings)]
+        (cond
+          (= (:type query) :query)
```
You could use `condp =` here

camsaul commented on src/metabase/query_processor/streaming.clj:
```diff
@@ -19,7 +19,8 @@
   (fn [initial-metadata]
-    (let [row-count (volatile! 0)]
+    (let [row-count                (volatile! 0)
+          ordered-col-viz-settings (:ordered-col-viz-settings initial-metadata)]
```
Is this used anywhere?

noahmoss replied:
Not yet. Will use it in the next PR to format XLSX exports. I can remove it from this PR.

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+          (= (:type query) :native)
+          (let [columns                  (-> viz-settings :table.columns)
+                ordered-col-viz-settings (update-native-query-viz-settings columns column-viz-settings)
+                rff' (fn [metadata] (rff (assoc metadata :ordered-col-viz-settings ordered-col-viz-settings)))]
+            (qp query rff' context))))
```
The `rff'` and `(qp query rff' context)` code is the same between the two branches here... it looks like only `fields`/`column` is different. I'd push the `cond`/condp =` down to that point so you don't need to duplicate all the other code.

noahmoss replied:
The function called to generate `ordered-col-viz-settings` is also different between the branches. But I think I see what you mean.

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
+
+(defn- get-viz-settings
```
Just call this `viz-settings`. I recommend reading https://stuartsierra.com/2016/01/09/how-to-name-clojure-functions, it does a great job of explaining that stuff

camsaul commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -0,0 +1,72 @@
+  the QP store (and defined in the data model settings), and are ordered based on the field list from the query.
+
+  For native queries, only viz settings passed from the frontend are incorporated.
+  TODO: native queries only pass viz settings if columns have been hidden or reordered. This needs to be fixed on the frontend."
```
I would add some documentation that mentions where this info is added (i.e. under the `:ordered-col-viz-settings` key in metadata) so we don't need to read the code to figure out where it gets added

camsaul commented on src/metabase/api/dataset.clj:
```diff
@@ -83,25 +84,37 @@
 
+(def ^:private column-ref-regex #"\[\"ref\",\[(?:\"field\",\d+,null|\"expression\",\".+\")\]\]")
+
+(defn- parse-viz-setting-key
```
It doesn't seem like this function actually parses anything??

noahmoss replied:
Yeah I was thinking of it as "the key function for parsing viz settings" but I can rename it

camsaul reviewed (commented):
I think this approach looks good to me in general. Ping me again when it has tests and I'll 👍 it

noahmoss committed:
squash commit to cleanup history

noahmoss committed:
tests

noahmoss committed:
big refactor to maintain hidden cols and col order for native query
exports

noahmoss committed:
allow output-order to be nil to skip ordering/filtering

noahmoss committed:
fix some tests

noahmoss committed:
fix more tests

noahmoss committed:
fix more tests

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
fix final (?) tests

noahmoss committed:
make helper private

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
fix cypress tests

noahmoss commented:
Changes to this PR since the first review:
* I changed the frontend to pass the stand-alone `visualizationSettings` prop in the API call, rather than the viz settings field on the `card` prop, which fixed the native query weirdness from the original PR description.
* I scrapped the column ordering logic from the viz settings middleware. Instead, I'm passing the full viz settings map in `metadata`, including the `:table.columns` key which contains ordering info. Then, in all the export streaming modules, I'm using correlating the names in `:table.columns` with the names in `cols` to determine the ordering of each row.
* The above changes mean that this is not a no-op PR. Instead, reordered and hidden columns should now be respected for native queries, bringing it to parity with MBQL.
* Added a full test suite for viz settings middleware.

noahmoss commented on src/metabase/query_processor/streaming/json.clj:
```diff
@@ -16,23 +16,32 @@
    :headers      {"Content-Disposition" (format "attachment; filename=\"query_result_%s.json\""
                                                 (u.date/format (t/zoned-date-time)))}})
 
+(defn- select-indices
```
I redefined this helper function in each streaming namespace because I wasn't sure where it should go. Is it general enough to go somewhere like `metabase.util`?

noahmoss committed:
add issue number to fix col deduping workaround

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
move around ordering logic

noahmoss committed:
fix tests

noahmoss committed:
fix remaining test

noahmoss committed:
preserve base field settings even when col doesn't have card-level settings

noahmoss committed:
Merge branch 'master' into export-formatting

camsaul reviewed (approved):
backend stuff LGTM.

I don't really understand what the reasoning is behind converting all the keys in the viz settings map to namespaced keywords but it looks like that was an existing pattern rather than something new to this PR.

noahmoss commented:
> I don't really understand what the reasoning is behind converting all the keys in the viz settings map to namespaced keywords but it looks like that was an existing pattern rather than something new to this PR.

Yeah, I'm following the existing format defined in `visualization_settings.cljc`. Maybe @jeff303 has context?

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
skip update-card-viz-settings if :fields is not in query

noahmoss committed:
add integer guard on field IDs to fix cypress test

noahmoss committed:
logic tweaks for some more edge cases

jeff303 commented:
> > I don't really understand what the reasoning is behind converting all the keys in the viz settings map to namespaced keywords but it looks like that was an existing pattern rather than something new to this PR.
> 
> Yeah, I'm following the existing format defined in `visualization_settings.cljc`. Maybe @jeff303 has context?

That was a judgement call I took based on the following:

- In general, namespaced keywords seem to be favored for new code (although of course there is debate on that)
- We added [spec](https://clojure.org/guides/spec) integration, and its idiomatic usage is for namespaced keywords (although of course it does support unqualified keys)
- A lot of the map keys were vague terms like `:name`.  Not only did they have different meanings even within viz settings (since that non-namespaced KW might appear in different kinds of maps), it also sometimes was ambiguous with that same keyword in other namespaces in the product.  I thought it would be nice to remove all the ambiguity and make sure every possible key had its own (more verbose, but more descriptive) key name.

noahmoss committed:
fix tests broken by last commit

noahmoss committed:
*actually* fix tests

noahmoss committed:
*actually* actually fix tests 😣
